### PR TITLE
WINC-1090: Support custom CA certificates for cluster-wide proxy

### PR DESCRIPTION
**Pull request by saifshaikh48**
Cluster admins can provide custom certificates that are required to communicate through their man-in-the-middle egress proxy. This PR covers importing certs Windows nodes so they can become fully proxy aware. 
This is done by updating the local trust store of all Windows instances with the certs present in the trusted-ca ConfigMap.

### CODE REVIEW

**Comment by saifshaikh48** [comment]
pull-ci-openshift-windows-machine-config-operator-release-4.10-vsphere-e2e-upgrade

**Comment by saifshaikh48** [comment]
/test vsphere-proxy-e2e-operator

cluster spinup failure

**Comment by saifshaikh48** [comment]
/test vsphere-proxy-e2e-operator

check after rebase

File: pkg/daemon/controller/controller.go [hunk 1]
```diff
@@ -282,15 +282,16 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
 	return ctrl.Result{}, nil
 }
 
-// reconcileEnvironmentVariables makes sure that the proxy variables exist as expected, or are safely rectified.
+// reconcileProxySettings makes sure that the proxy variables exist as expected, or are safely rectified.
```
Comment (aravindhp) [inline pkg/daemon/controller/controller.go]: I am not sure about this change. WICD should be generic wrt environment variables. It shouldn't be proxy specific. These variables just so happen to be related to enabling proxy in the environment.

File: pkg/daemon/certs/certs.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+// Reconcile ensures the certificates in the given trusted CA bundle file are imported as system certificates.
+// Returns a boolean whether any certs have needed to be imported
+func Reconcile(trustBundle string) (bool, error) {
+	if trustBundle == "" {
+		// TODO: ensure that all certs that we've added previously are removed https://issues.redhat.com/browse/WINC-998
+		return false, nil
+	}
+	// Read expected certs from CA trust bundle file
+	pemData, err := ioutil.ReadFile(trustBundle)
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:24]: Isn't `ioutil.ReadFile` deprecated? In general we shouldn't we trying to read the whole file in one shot. There could be potential of a DOS. It might be better to use `os.Open()` and then `io.Copy()`. Using a context aware reader will also help.

File: pkg/daemon/certs/certs.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	"syscall"
+	"unsafe"
+
+	syswindows "golang.org/x/sys/windows"
+	"k8s.io/klog/v2"
+)
+
+// Reconcile ensures the certificates in the given trusted CA bundle file are imported as system certificates.
+// Returns a boolean whether any certs have needed to be imported
+func Reconcile(trustBundle string) (bool, error) {
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:18]: Can this function be made readable? Can you have a trust bundle object that you act upon?

**Review by aravindhp — COMMENTED** [review]
@saifshaikh48 just left some early feedback.

File: pkg/daemon/certs/certs.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	if trustBundle == "" {
+		// TODO: ensure that all certs that we've added previously are removed https://issues.redhat.com/browse/WINC-998
```
Comment (mansikulkarni96) [inline pkg/daemon/certs/certs.go:20]: WINC-688*

File: pkg/daemon/certs/certs.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
+package certs
```
Comment (mansikulkarni96) [inline pkg/daemon/certs/certs.go:1]: Can you add the copyright text headers here.

File: pkg/daemon/certs/certs.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+			}
+			// Add the certificate to the instances's system-wide trust store
+			if err = syswindows.CertAddCertificateContextToStore(store, certContext,
+				syswindows.CERT_STORE_ADD_REPLACE_EXISTING, nil); err != nil {
+				klog.Errorf("Failed to import certificate %s: %v\n", cert.Subject.String(), err)
+				continue
+			}
+			certChange = true
+			klog.Infof("Certificate %s imported successfully.", cert.Subject.String())
+		}
```
Comment (mansikulkarni96) [inline pkg/daemon/certs/certs.go:69]: will we retry the cert import if we don't return errors?

File: pkg/daemon/controller/controller.go (hunk 1 above)
Comment (mansikulkarni96) [reply pkg/daemon/controller/controller.go]: I see that you are using this function to reconcile certs in the next commit, if so it can be named along the lines of
reconcileEnvVarsAndCerts or you could create another function that calls both of these.

File: pkg/daemon/certs/certs.go (hunk 6 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:69]: yes, it'll retry any failed ones on the next WICD controller Reconcile() call (every 30 seconds)

File: pkg/daemon/certs/certs.go (hunk 2 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:24]: Yup, it's deprecated, I'll use os.Open() and then io.Copy() instead.

I also can explore reading in the file one cert at a time, maybe using the `-----END CERTIFICATE-----` as a delimiter to the buffered reader

File: pkg/daemon/controller/controller.go (hunk 1 above)
Comment (saifshaikh48) [reply pkg/daemon/controller/controller.go]: I'll rename the func and squash this commit with the next so it's clear how it'll be used

File: pkg/daemon/certs/certs.go (hunk 3 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:38]: need to explore this more

Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:40]: Introduced more helper functions to keep this top level func more readable

File: pkg/daemon/certs/certs.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+	}
+
+	// Open the root certificate store
+	systemStore, err := windows.CertOpenStore(windows.CERT_STORE_PROV_SYSTEM, 0, 0,
+		windows.CERT_SYSTEM_STORE_LOCAL_MACHINE, uintptr(unsafe.Pointer(syscall.StringToUTF16Ptr("ROOT"))))
+	if err != nil {
+		return false, fmt.Errorf("Failed to open root certificate store: %v", err)
+	}
+	defer windows.CertCloseStore(systemStore, 0)
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:57]: Please don't ignore the error. Printing it will suffice.

File: pkg/daemon/certs/certs.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	return ensureCertsAreUpToDate(systemStore, expectedCerts, existingCerts), nil
+}
+
+// readCertsFromFile reads in any PEM-encoded certificates from the given file
+func readCertsFromFile(path string) ([]*x509.Certificate, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open file: %v", err)
+	}
+	defer f.Close()
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:74]: Same comment here.

File: pkg/daemon/certs/certs.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+				break
+			}
+			return nil, fmt.Errorf("Error reading existing system certificate: %v", err)
+		}
+
+		// Convert the certificate bytes to Golang x509.Certificate
+		certBytes := unsafe.Slice(certContext.EncodedCert, certContext.Length)
+		x509Cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			klog.Errorf("Failed to parse certificate from bytes %s: %v", certBytes, err)
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:144]: This implies we are ignoring the error. We need to guarantee that all certs are handled and errors are returned even if one fails.

File: pkg/daemon/certs/certs.go [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+		}
+		existingSystemCerts = append(existingSystemCerts, x509Cert)
+	}
+	klog.Infof("found %d existing system certificates", len(existingSystemCerts))
+	return existingSystemCerts, nil
+}
+
+// ensureCertsAreUpToDate imports any of the expected certificates that are not already present in the existing list
+// into the given Windows certificate store. Returns a boolean if any certificates were imported
+func ensureCertsAreUpToDate(store windows.Handle, expected, existing []*x509.Certificate) bool {
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:155]: `ensureCertsAreImported`?

File: pkg/daemon/controller/controller.go [hunk 11]
```diff
@@ -282,15 +283,20 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
...
-func (sc *ServiceController) reconcileEnvironmentVariables(envVars map[string]string, watchedEnvVars []string,
+func (sc *ServiceController) reconcileEnvVarsAndCerts(envVars map[string]string, watchedEnvVars []string,
 	node core.Node) (bool, error) {
-	restartRequired, err := envvar.EnsureVarsAreUpToDate(envVars, watchedEnvVars)
+	certsUpdated, err := certs.Reconcile(sc.trustBundle)
 	if err != nil {
 		return false, err
 	}
-	if restartRequired {
+	envVarsUpdated, err := envvar.EnsureVarsAreUpToDate(envVars, watchedEnvVars)
```
Comment (aravindhp) [inline pkg/daemon/controller/controller.go:294]: We should rename the function to `envvar.Reconcile` in a follow up commit.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @saifshaikh48. Please address my comment.s

File: pkg/daemon/certs/certs.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+	if err != nil {
+		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+	}
+
+	return ensureCertsAreUpToDate(systemStore, expectedCerts, existingCerts), nil
+}
+
+// readCertsFromFile reads in any PEM-encoded certificates from the given file
+func readCertsFromFile(path string) ([]*x509.Certificate, error) {
+	f, err := os.Open(path)
```
Comment (sebsoto) [inline pkg/daemon/certs/certs.go:70]: I suggest using os.ReadFile() here.
The use of a reader for non streaming data is making this function a lot more complicated than it needs to be.
Readers are very useful for pipes and extremely large files, but in the case of a file that we know won't cause any memory issues, just use ReadFile and string parsing.

```
os.ReadFile()
strings.Split(endCertTag)
// for each:
strings.TrimSpace
```

File: pkg/daemon/certs/certs.go (hunk 9 above)
Comment (sebsoto) [reply pkg/daemon/certs/certs.go:144]: +1, I don't see a reason to not return an error here.

File: pkg/daemon/certs/certs.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+// into the given Windows certificate store. Returns a boolean if any certificates were imported
+func ensureCertsAreUpToDate(store windows.Handle, expected, existing []*x509.Certificate) bool {
+	certChange := false
+	for _, cert := range expected {
+		if containsCert(existing, cert) {
+			// Cert already exists as expected, do nothing
+			continue
+		}
+
+		// Convert x509 certificate to a Windows CertContent
```
Comment (sebsoto) [inline pkg/daemon/certs/certs.go:163]: CertContent -> CertContext ?

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/daemon/certs/certs.go (hunk 12 above)
Comment (sebsoto) [reply pkg/daemon/certs/certs.go:70]: This will remove the need for the splitAt() function being introduced

File: pkg/daemon/certs/certs.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+		return false, nil
+	}
+	// Read expected certs from CA trust bundle file
+	expectedCerts, err := readCertsFromFile(trustBundle)
+	if err != nil {
+		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+	}
+
+	// Open the root certificate store
+	systemStore, err := windows.CertOpenStore(windows.CERT_STORE_PROV_SYSTEM, 0, 0,
```
Comment (jrvaldes) [inline pkg/daemon/certs/certs.go:52]: Does it make sense to cache the `store` handler so that it get initialized once during the lifespan of the ServiceController?

**Review by jrvaldes — COMMENTED** [review]
Good work @saifshaikh48, left a few comments PTAL.

File: pkg/windows/windows.go [hunk 15]
```diff
@@ -470,7 +471,9 @@ func (vm *windows) ConfigureWICD(watchNamespace, wicdKubeconfigContents string)
 	}
 	wicdServiceArgs := fmt.Sprintf("controller --windows-service --log-dir %s --kubeconfig %s --namespace %s",
 		wicdLogDir, wicdKubeconfigPath, watchNamespace)
-
+	if cluster.IsProxyEnabled() {
+		wicdServiceArgs = fmt.Sprintf("%s --trust-bundle %s", wicdServiceArgs, TrustedCABundlePath)
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/windows/windows.go:475]: Yes, CNO decides what certs to inject into the trusted-ca configmap in all proxied environments
So even if only HTTP is used and even if the trust bundle is empty, the file will still exist.

File: pkg/daemon/certs/certs.go (hunk 12 above)
Comment (aravindhp) [reply pkg/daemon/certs/certs.go:70]: @saifshaikh48 had used `os.ReadFile()` originally and I had requested the change to not use it. I am not a fan of using `os.ReadFile()` on files whose content we have no control over.

File: pkg/daemon/certs/certs.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:50]: Do you see a case where the deferred close doesn't work as intended?

File: pkg/daemon/certs/certs.go (hunk 12 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:72]: I was initially reading the file directly into memory but @aravindhp brought up a [good point](https://github.com/openshift/windows-machine-config-operator/pull/1707#discussion_r1305338337). Let me know if you feel strongly that the reader is not needed

The cert bundle has the potential to be large, since it is a combination of user provided certs and cluster-wide trusted CA certificates for platform components ([as deduced by CNO](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.12/html/networking/configuring-a-custom-pki))

File: pkg/daemon/certs/certs.go [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+		if err != nil {
+			klog.Errorf("Failed to create content from x509 cert %s: %v\n", cert.Subject.String(), err)
+			continue
+		}
+		// Add the certificate to the instance's system-wide trust store
+		if err = windows.CertAddCertificateContextToStore(store, certContext,
+			windows.CERT_STORE_ADD_REPLACE_EXISTING, nil); err != nil {
+			klog.Errorf("Failed to import certificate %s: %v\n", cert.Subject.String(), err)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go]: I was thinking that we shouldn't stop reconciling the rest of the certs if one fails, but I've been convinced otherwise

File: pkg/daemon/certs/certs.go [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+			return 0, nil, nil
+		}
+
+		// Find next separator and return the token
+		if i := bytes.Index(data, searchBytes); i >= 0 {
+			return i + searchLen, data[0 : i+searchLen], nil
+		}
+
+		if atEOF {
+			// Hit end of file before finding another separator, return rest of the data
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go]: good, point will return an error

File: pkg/daemon/certs/certs.go [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+			klog.Errorf("Failed to parse certificate: %v\n", err)
+			continue
+		}
+		certs = append(certs, cert)
+	}
+	return certs, nil
+}
+
+// splitAt is a custom closure function to split data into complete PEM-encoded certificates
+func splitAt(substring string) func(data []byte, atEOF bool) (advance int, token []byte, err error) {
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:105]: not intended, fixed

File: pkg/daemon/certs/certs.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply pkg/daemon/certs/certs.go:50]: I don't, this is something preventive. I recognize this is a followed pattern so I don't want to block the PR on it.
This can be proposed in a follow-up PR and we can get it updated everywhere if needed.

File: pkg/daemon/certs/certs.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:105]: need to figure out the best way to unit test this, TODO

File: test/e2e/proxy_test.go [hunk 19]
```diff
@@ -32,6 +34,44 @@ func proxyTestSuite(t *testing.T) {
 
 	t.Run("Trusted CA ConfigMap validation", testTrustedCAConfigMap)
 	t.Run("Environment variables validation", testEnvVars)
+	t.Run("Certificate validation", testCerts)
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply test/e2e/proxy_test.go:37]: I looked into this a bit but am unsure what endpoint to use with the probe

File: pkg/daemon/certs/certs.go [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+			return 0, nil, nil
+		}
+
+		// Find next separator and return the token
+		if i := bytes.Index(data, searchBytes); i >= 0 {
+			return i + searchLen, data[0 : i+searchLen], nil
+		}
+
+		if atEOF {
+			return dataLen, data, fmt.Errorf("Hit end of file without terminating the current token")
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:121]: Some context here too will help.

File: pkg/daemon/certs/certs.go [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+				break
+			}
+			return nil, fmt.Errorf("Error reading existing system certificate: %v", err)
+		}
+
+		// Convert the certificate bytes to Golang x509.Certificate
+		certBytes := unsafe.Slice(certContext.EncodedCert, certContext.Length)
+		x509Cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate from bytes %s: %v", certBytes, err)
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:148]: Is `certBytes` sensitive?

File: pkg/daemon/certs/certs.go [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+		if err := windows.CertCloseStore(systemStore, 0); err != nil {
+			klog.Error("Failed to close root system certificate store")
+		}
+	}()
+
+	// Get all existing from system store
+	existingCerts, err := getAllCerts(systemStore)
+	if err != nil {
+		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+	}
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:65]: This code should ideally be in `ensureCertsAreImported()`. 

PS: Even better have a struct around all of this but I won't insist. If I have some cycles I will try to do a refactor just to show an example.

File: pkg/daemon/certs/certs.go [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+		if containsCert(existing, cert) {
+			// Cert already exists as expected, do nothing
+			continue
+		}
+
+		// Convert x509 certificate to a Windows CertContext
+		certContext, err := windows.CertCreateCertificateContext(
+			windows.X509_ASN_ENCODING|windows.PKCS_7_ASN_ENCODING, &cert.Raw[0], uint32(len(cert.Raw)))
+		if err != nil {
+			return false, fmt.Errorf("Failed to create content from x509 cert %s: %v\n", cert.Subject.String(), err)
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:170]: Is the subject always populated?

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/daemon/certs/certs.go (hunk 12 above)
Comment (sebsoto) [reply pkg/daemon/certs/certs.go:72]: Thats fine with me.

File: test/e2e/proxy_test.go [hunk 24]
```diff
@@ -32,6 +34,44 @@ func proxyTestSuite(t *testing.T) {
...
+}
+
+func testCerts(t *testing.T) {
+	tc, err := NewTestContext()
+	require.NoError(t, err)
```
Comment (sebsoto) [inline test/e2e/proxy_test.go:42]: I missed this in previous proxy PRs, but this pattern of creating a new testContext within each individual test has been removed from the rest of the e2e tests. Instead NewTestContext is called once at the beginning of each test suite, and the tests which require its use are made methods of it.

The reason for this ask is that by using the previous pattern, each test ends up having a repeated boiler plate code within them. 

If you can please add a commit changing this for the rest of the tests. If you rather not do that within this PR I am fine with it being done only for this test. Meaning within this PR changing only this function to be a method of testContext, without changing the others.

File: test/e2e/proxy_test.go [hunk 25]
```diff
@@ -32,6 +34,44 @@ func proxyTestSuite(t *testing.T) {
...
+
+	// Read all certs from CM data
+	trustedCABundle := cm.Data[certificates.CABundleKey]
+	assert.Greater(t, len(trustedCABundle), 0, "no additional user-provided certs in bundle")
+
+	certs := x509.NewCertPool()
+	require.True(t, certs.AppendCertsFromPEM([]byte(trustedCABundle)), "unable to parse certs from trusted CA ConfigMap data")
+	subjects := certs.Subjects()
+	// Ensure each cert has been imported into every Windows instance's system store
+	for _, node := range gc.allNodes() {
```
Comment (sebsoto) [inline test/e2e/proxy_test.go:56]: Where is `gc.allNodes` being set?
What happens if a developer just wants to run this individual test?

File: pkg/daemon/certs/certs.go (hunk 14 above)
Comment (sebsoto) [reply pkg/daemon/certs/certs.go:50]: Its safer to only open the handle when we specifically need it.

File: pkg/daemon/certs/certs.go [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			klog.Errorf("Failed to close file %s", f.Name())
+		}
+	}()
+
+	// Process the file one certificate at a time
+	scanner := bufio.NewScanner(f)
+	const endCertTag = "-----END CERTIFICATE-----"
```
Comment (sebsoto) [inline pkg/daemon/certs/certs.go:84]: nit: this could actually be moved within `splitAt`, and its parameter removed, unless theres a need for it to be a generic function :sweat_smile:

File: pkg/daemon/certs/certs.go (hunk 21 above)
Comment (sebsoto) [reply pkg/daemon/certs/certs.go:148]: +1

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/daemon/certs/certs.go (hunk 12 above)
Comment (sebsoto) [reply pkg/daemon/certs/certs.go:72]: I guess it could be an attack vector. :thinking:

File: pkg/daemon/certs/certs.go (hunk 14 above)
Comment (jrvaldes) [reply pkg/daemon/certs/certs.go:50]: True

File: pkg/daemon/certs/certs.go (hunk 23 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:170]: It'll always be non-nil. Worst case Subject.String() will return empty string so no worry of panic

File: pkg/daemon/certs/certs.go (hunk 21 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:148]: potentially, I'll remove it from the error msg

File: pkg/daemon/certs/certs.go [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+	scanner.Split(splitAt(endCertTag))
+	var certs []*x509.Certificate
+	for scanner.Scan() {
+		// Read next token
+		pemData := scanner.Text()
+		block, _ := pem.Decode([]byte(pemData))
+		if block == nil {
+			return nil, fmt.Errorf("No PEM data read from token")
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:92]: It could leak a cert -- IMO best to be cautious and just let the user know the issue is that the file data is not PEM-encoded

File: test/e2e/proxy_test.go (hunk 25 above)
Comment (saifshaikh48) [reply test/e2e/proxy_test.go:56]: I see, I'll call `tc.loadExistingNodes()` at the start of the proxy test suite so we can use gc.allNodes() safely

We might need to do the same in the storage and network test suites, could you confirm my understanding? If so that can be done in a followup PR

File: test/e2e/proxy_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (mansikulkarni96) [reply test/e2e/proxy_test.go:37]: That depends on the certs we are using, can you remind me where the certs are created for the test suite, is it through the job itself?

[the remark this answers is not in the document]
Comment (saifshaikh48) [reply test/e2e/proxy_test.go]: For the test suite, the additional certs are the RHCOS trust bundle (CNO provides this to us via the trusted-ca configmap)

File: pkg/daemon/certs/certs.go [hunk 28]
```diff
@@ -0,0 +1,191 @@
...
+			return nil, fmt.Errorf("Error reading existing system certificate: %v", err)
+		}
+
+		// Convert the certificate bytes to Golang x509.Certificate
+		certBytes := unsafe.Slice(certContext.EncodedCert, certContext.Length)
+		x509Cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate from bytes: %v", err)
+		}
+		existingSystemCerts = append(existingSystemCerts, x509Cert)
```
Comment (aravindhp) [inline pkg/daemon/certs/certs.go:177]: Nit: this would in turn become a helper method: `addCertFromStore()`

**Review by aravindhp — COMMENTED** [review]
LGTM. I will let @sebsoto approve as he had asked for changes.

File: test/e2e/proxy_test.go (hunk 25 above)
Comment (sebsoto) [reply test/e2e/proxy_test.go:59]: yeah either tc.loadExistingNodes() or waitForWindowsNodes needs to be called. Things could be a bit more consistent, but both achieve the same thing of populating the nodes within gc.

**Review by sebsoto — APPROVED** [review]

File: pkg/daemon/certs/certs.go [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"k8s.io/klog/v2"
+)
+
+// Reconcile ensures the certificates in the given trusted CA bundle file are imported as system certificates.
+// Returns a boolean whether any certs needed to be reconciled
+func Reconcile(caBundle string) (bool, error) {
```
Comment (jrvaldes) [inline pkg/daemon/certs/certs.go:38]: s/caBundle/caBundlePAth

to avoid confusion with the ca bundle raw data

File: cmd/daemon/controller.go [hunk 30]
```diff
@@ -39,6 +39,7 @@ var (
 	}
 	windowsService bool
 	logDir         string
+	caBundle       string
```
Comment (jrvaldes) [inline cmd/daemon/controller.go:42]: s/caBundle/caBundlePath

File: pkg/daemon/certs/certs.go [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+
+// readCertsFromFile reads in any PEM-encoded certificates from the given file
+func readCertsFromFile(path string) ([]*x509.Certificate, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open file: %v", err)
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			klog.Errorf("Failed to close file %s", f.Name())
```
Comment (jrvaldes) [inline pkg/daemon/certs/certs.go:59]: In case of an issue closing the handler  the error is getting swallowed by the log, should this thrown instead?

File: pkg/daemon/certs/certs.go [hunk 32]
```diff
@@ -0,0 +1,198 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %v\n", err)
+		}
+		certs = append(certs, cert)
+	}
+	return certs, nil
+}
+
+// splitAtPEMCert is a custom closure function to split data into complete PEM-encoded certificates
+func splitAtPEMCert() func(data []byte, atEOF bool) (advance int, token []byte, err error) {
```
Comment (jrvaldes) [inline pkg/daemon/certs/certs.go:85]: It would be nice to cover this with unit tests; especially for the edge cases i.e. `if atEOF && dataLen == 0` , `if atEOF`, etc

File: pkg/daemon/certs/certs.go [hunk 33]
```diff
@@ -0,0 +1,198 @@
...
+		certChange = true
+	}
+	return certChange, nil
+}
+
+// getAllCerts reads all the certs from the given certificate store
+func getAllCerts(store windows.Handle) ([]*x509.Certificate, error) {
+	var existingSystemCerts []*x509.Certificate
+	var certContext *windows.CertContext
+	var err error
```
Comment (jrvaldes) [inline pkg/daemon/certs/certs.go:150]: Consider declaring the error and the list as return parameters in the function's signature.

**Review by jrvaldes — APPROVED** [review]
LGTM, added minor suggestions.

File: pkg/daemon/certs/certs.go (hunk 32 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:85]: True, I tried to think through the best way to unit test but couldn't write something clean. Maybe I'll write unit tests for the func that uses this `readCertsFromFile` instead (followup PR)

File: pkg/daemon/certs/certs.go (hunk 31 above)
Comment (saifshaikh48) [reply pkg/daemon/certs/certs.go:59]: throwing the err in this defer block would overwrite any other error returned by the func. Logging such errors is a compromise between that and ignoring them completely

**Comment by saifshaikh48** [comment]
/test vsphere-proxy-e2e-operator

Deletion test failure
